Paper: Papyri: Better documentation for the Scientific Ecosystem in Jupyter

Carreau opened the pull request:
See http://procbuild.scipy.org/ for logs generated by the build process.

Thanks for all your work on organising SciPy.



*Editor*: <!--editor-->@stargaser<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@wd15, @karthikmurugadoss<!--end-reviewers-list-->

Carreau committed:
start

carvalhocamille committed:
edit papyri.rst test

Carreau committed:
Merge pull request #1 from carvalhocamille/2022

edit papyri.rst test

Carreau committed:
updates

Carreau committed:
updates

Carreau committed:
misc

Carreau committed:
Merge remote-tracking branch 'github/2022' into 2022

carvalhocamille committed:
small edits

carvalhocamille committed:
edits rst intro/motiv

Carreau committed:
Merge pull request #2 from carvalhocamille/2022

small edits abstract

Carreau committed:
Merge pull request #3 from carvalhocamille/edits051022

Edits051022

Carreau committed:
syn

Carreau committed:
Merge remote-tracking branch 'github/2022' into 2022

Carreau committed:
add ping

Carreau committed:
update

Carreau committed:
more

carvalhocamille committed:
edits intro

carvalhocamille committed:
edits intro and section 1 (need 1.3 to be edited still)

Carreau committed:
Merge pull request #4 from carvalhocamille/editsCC0516

Edits cc0516

Carreau committed:
typoes

carvalhocamille committed:
edits section 1.3

Carreau committed:
Merge pull request #6 from carvalhocamille/section1.3_bis

edits section 1.3

Carreau committed:
precision

Carreau committed:
table pblm

Carreau committed:
more updates

Carreau committed:
more updates

carvalhocamille committed:
edits papyri solution part

carvalhocamille committed:
new edits on IRD files

Carreau committed:
Merge pull request #7 from carvalhocamille/edits_train_0522

edits papyri solution part

Carreau committed:
syntax

Carreau committed:
Minor rephrase

Carreau committed:
some updates

Carreau committed:
add file

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

deniederhut commented:
@scoobies mark paper

deniederhut commented:
@scoobies build status

scoobies commented:
{"Carreau-2022"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpza_d6ary'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: bussonnier\n*** Warning: /tmp/tmpza_d6ary/publisher/../output/bussonnier/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/Carreau-2022.pdf", "build_status"=>"success", "build_timestamp"=>"27/05 20:19"}, "status"=>"success"}}

deniederhut commented:
@scoobies mark needs review

deniederhut commented:
@scoobies assign @stargaser as editor

scoobies commented:
Assigned! @stargaser is now the editor

carvalhocamille committed:
edit sections1-3 CC

Carreau committed:
Merge pull request #8 from carvalhocamille/edits_CC_6122

edit sections1-3 CC

wd15 commented:
Hi, I've been asked to review this. Looking forward to it. Deadline for my input in June 9.

stargaser commented:
@scoobies add @wd15 to reviewers

scoobies commented:
@wd15 added to the reviewers list!

deniederhut commented:
@scoobies add @karthikmurugadoss to reviewers

scoobies commented:
@karthikmurugadoss added to the reviewers list!

carvalhocamille committed:
end edits section 3. Fixed typos, reformulated sentences.

Carreau committed:
Update papers/bussonnier/papyri.rst

Carreau committed:
Merge pull request #9 from carvalhocamille/edits_CC_6122

end edits section 3. Fixed typos, reformulated sentences.

carvalhocamille committed:
last edits

Carreau committed:
Update papers/bussonnier/papyri.rst

Carreau committed:
Merge pull request #10 from carvalhocamille/last_edits_CC

last edits

deniederhut commented:
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

wd15 commented:
Papyri is addressing a real need for improved documentation procedure
in scientific open-source development. A model that separates the
generation of API / example documentation from the view of the
document would be a major step forward in standardizing documentation
procedure across IDEs, web views and applications. This will greatly
reduce the document maintenance overhead required when developing
open-source software.

Overall, the paper gives a good introduction into the thinking behind
Papyri. These are my initial thoughts on how to improve the paper.

 - The grammar seems a bit loose in places. I can go through as best I
   can and mark where I notice grammar issues, but I suggest the
   authors give this another go or maybe pass it through a grammar
   engine. I'll look through it again once you've given it another
   read through.

 - The intermediate format or IRD is a very important step for the
   community. Other tools can build from this format either by
   generating the documentation view or by generating the IRD from the
   source code. It would be nice in the paper if the authors could
   actually describe the details of the format or schema for the
   IRD. The schema itself could become a standard for documentation
   and, thus, making it transparent to the reader would be useful.

 - Suggestion: Pandoc is a tool that uses an AST and can covert
   between many markup and documentation formats. Would it be useful
   to mention Pandoc in the paper as an example of a successful tool
   that uses a similar approach?

 - Suggestion: As part of the IRD, package metadata is stored. Would
   it be useful to use an existing schema such as [CODEMETA.yaml](https://codemeta.github.io/) for
   this.

 - Section numbering could be improved by using "X.Y" for
   subsections. Also, shouldn't the Introduction be numbered as Section 1?
 
I'll give you a chance to clean up the paper and give a more detailed
line-by-line review over the next week.

deniederhut commented:
@scoobies mark needs more review

carvalhocamille committed:
full read and typos grammar edits

karthikmurugadoss commented:
The paper presented here discusses Papyri which is an approach for unifying the documentation experience in the scientific Python ecosystem. Primarily, Papyri aims at decoupling the process of creating documentation (using the presented IRD format) and actually rendering the documentation (which can often times be user specific)

The paper describes the motivation and current state of Papyri quite well. The following are my major and minor comments. 

Major Comments:
1. Section on current implementation conveys a lot of information and is a bit difficult to follow for someone who isn't well-versed with existing documentation workflows. A few concepts that are introduced are RST parsing, CBOR representation, etc. 
2. The section on `IRD file installation` can benefit from a visual showing the different components (SQLite database, Raw storage on disk, etc.) and how these components interact with each other. 
3. Related to the above point, it is not clear why the object information is stored in 3 different places and what exactly is stored in each of these locations. More clarity here would be very helpful.
4. The context and usefulness of the local graph visualization is not described. Does Papyri create this visual as a part of the documentation generation process? Or is it here specifically to highly the connections into/from `ndarray`? 

Minor Comments:
1. There are typos and missing words in a number of places which need to be addressed. 
2. In the Current Implementation section, there are a number of referenced tools for which links to their sources would be helpful. e.g. Jedi, Pygments, Quart, Trio, etc. 
3. The numbering for sections can be improved

deniederhut commented:
@scoobies mark pending comment

Carreau committed:
Merge pull request #11 from carvalhocamille/CC_typos_edits

full read and typos grammar edits

Carreau committed:
Response to comments:

from wd15:

> The intermediate format or IRD is a very important step for the
> community. Other tools can build from this format either by
> generating the documentation view or by generating the IRD from the
> source code. It would be nice in the paper if the authors could
> actually describe the details of the format or schema for the
> IRD. The schema itself could become a standard for documentation
> and, thus, making it transparent to the reader would be useful.

We've extended the paragraph speaking of this. As the IRD isstill
changing rapidly we don't belove a description that woudl be outdated in
a week would be useful in the paper.

Carreau committed:
Address comment:

from wd15:

> The intermediate format or IRD is a very important step for the
> community. Other tools can build from this format either by
> generating the documentation view or by generating the IRD from the
> source code. It would be nice in the paper if the authors could
> actually describe the details of the format or schema for the
> IRD. The schema itself could become a standard for documentation
> and, thus, making it transparent to the reader would be useful.

We want to avoid duplication, and would prefer to point to other
medatada source. Currently we limit to only what is necessary

Carreau committed:
rework section on IRD file installation

Carreau committed:
major rework on current implementation section

Carreau committed:
Extend on network graph

Carreau committed:
Add multiple reference reference and citations

Carreau committed:
finish citation

Carreau commented:
Many thanks for both reviews, I've tried to address most of the above points in different commits to try to make re-review easier. 

Here is a small summary of the changes.

> The intermediate format or IRD is a very important step for the
> community. Other tools can build from this format either by
> generating the documentation view or by generating the IRD from the
> source code. It would be nice in the paper if the authors could
> actually describe the details of the format or schema for the
> IRD. The schema itself could become a standard for documentation
> and, thus, making it transparent to the reader would be useful.


I believe the project is too young to give a complete description of the IRD,
there are still regular changes to the format every 2 to 3 weeks depending on
the activity, thus a detail description would be premature. The IRD is still
changing much less frequently than initially but still too frequently IMHO. 
I tied to clarify this.


> Suggestion: Pandoc is a tool that uses an AST and can covert
> between many markup and documentation formats. Would it be useful
> to mention Pandoc in the paper as an example of a successful tool
> that uses a similar approach?

Yes, this is a good tool, I've added it. I was also recently made aware of
https://markdoc.io/docs/nodes, that was released recently but haven't had a
change to try it, so didn't wanted to make major changes that late in the
proceeding process. 

> Suggestion: As part of the IRD, package metadata is stored. Would
> it be useful to use an existing schema such as CODEMETA.yaml for
> this.

I've extended a bit this section, my main take so far is to limit to the minimum
vital, and rely on metadata that is stored somewhere else. I would much prefer
something like codemeta or other JSON LD format to be part of the package on
PyPI. 

> Section numbering could be improved by using "X.Y" for
> subsections. Also, shouldn't the Introduction be numbered as Section 1?

We removed section numbering altogether, it should be something that is handled
at the proceeding level. Is the proceeding is compiled with a directive like
`sectnum` then sections will have number, if we add them manually then the
number will appear twice.



> Section on current implementation conveys a lot of information and is a bit
> difficult to follow for someone who isn't well-versed with existing
> documentation workflows. A few concepts that are introduced are RST parsing,
> CBOR representation, etc.


I've extend where possible, but as for other comments above about IRD schema, I
don't want to dive too far into the implementation as it is still changing
quickly and should not be necessary to understand the idea.

I hope that in the future other competing projects will for that
produce/consume IRD bundles, and potentially make completely different technical
choices.

> The section on IRD file installation can benefit from a visual showing the
> different components (SQLite database, Raw storage on disk, etc.) and how
> these components interact with each other.

I've tried to clarify as well, and made a schema. I've also extended that these
choices are mostly made due to the current use case I target and could/should be
reconsider by an implementation with different targets.

> Related to the above point, it is not clear why the object information is
> stored in 3 different places and what exactly is stored in each of these
> locations. More clarity here would be very helpful.

As for above, I've tried to clarify, let me know if this is clearer.

> The context and usefulness of the local graph visualization is not described.
> Does Papyri create this visual as a part of the documentation generation
> process? Or is it here specifically to highly the connections into/from
> ndarray?


Thanks, this was indeed not clear, I've reworked this section. This was trying
to  demonstrate that changes to the documentation could be done without having
to re-do the generation step. This graph is indeed generated at render time, and
updates depending on which libraries you have documentation installed for, and
give you ideas of types of UI changes that could be implemented later. 


> There are typos and missing words in a number of places which need to be addressed.

We tried to fix things the best we could, and would appreciate any pointers to
remaining mistakes.

> In the Current Implementation section, there are a number of referenced tools for which links to their sources would be helpful. e.g. Jedi, Pygments, Quart, Trio, etc.


This should be mostly fixed, beyond a couple of citation I need to expand with
the right DOIs.

carvalhocamille committed:
full check

Carreau committed:
Merge pull request #12 from carvalhocamille/last_spell_check

full check

deniederhut commented:
Awesome! @wd15 and @karthikmurugadoss -- do you feel this paper is now ready for inclusion in the proceedings?

karthikmurugadoss commented:
Yes did another read through and it looks good to me!

Carreau commented:
> Yes did another read through and it looks good to me!

Many thanks, that's giving me extra motivation after the week-end !

wd15 commented:
Looks much better. Nice work!

stargaser commented on README.md:
```diff
@@ -477,6 +477,7 @@ Editors should come to a final 'ready', 'unready' decision before the **Final Ed
 5. Choose the appropriate folder, and make a new directory inside it (it needs a unique name)
 6. Copy your slide deck or poster into the directory, and add a file called `info.json` with the following fields needed for uploading to Zenodo (using an empty string for author orcid or
 affiliation if these cannot be provided):
+
```
@Carreau would you please remove the edit to the top-level README.md file? Files outside the papers directory cannot be modified.

Carreau replied:
Oops, sorry about that. Should be removed now. Should I rebase to make sure it is not even touched in the history ?

stargaser replied:
Thanks @Carreau, I think there is no need to rebase.

Carreau commented on README.md:
```diff
@@ -477,6 +477,7 @@ Editors should come to a final 'ready', 'unready' decision before the **Final Ed
 5. Choose the appropriate folder, and make a new directory inside it (it needs a unique name)
 6. Copy your slide deck or poster into the directory, and add a file called `info.json` with the following fields needed for uploading to Zenodo (using an empty string for author orcid or
 affiliation if these cannot be provided):
+
```
```suggestion
```

Carreau committed:
Update README.md

stargaser commented:
@scoobies mark ready
